### PR TITLE
Marketplace login

### LINKED ISSUE #425: Get API-Key via Marketplace instead of copying manually
The current system with api-key confuses a lot of people so we want to simplify the process by enabling an login from inside the market app. 

UX-Flow:
- User clicks "Login to marketplace" inside the market app
- New browser-tab opens with marketplace login form and authorization request
- User enters his marketplace credentials thus authorizing the market app to get the api-key via one-time code.

Note: This is not OAuth2, but a  similar custom flow, because OAuth2 would require static redirect-urls.

### PR DESCRIPTION
**Pull request by IljaN**
Login directly from market-app to auto-install your api-key. 

- Market app send its callback-url and [PKCE](https://tools.ietf.org/html/rfc7636)-Challenge to marketplace.
- A marketplace login and approval prompt is presented to the user.
- If user approves prompt, a one-time token is generated and forwarded to the market app.
- The Token is used together with PKCE-Verify-code by market app to retrieve the api-key.

This PR can be tested against staging marketplace by setting appstoreurl.

This is *NOT* OAuth because OAuth requires a static callback-url

Closes #425

### CODE REVIEW

File: lib/Controller/MarketController.php [hunk 1]
```diff
@@ -21,15 +22,20 @@
...
 use OCP\AppFramework\Controller;
 use OCP\AppFramework\Http;
 use OCP\AppFramework\Http\DataResponse;
+use OCP\IAppConfig;
 use OCP\IL10N;
 use OCP\IRequest;
 use OCP\IConfig;
+use OCP\IURLGenerator;
+use OCP\Security\ISecureRandom;
+use Punic\Exception;
```
Comment (VicDeo) [inline lib/Controller/MarketController.php]: unused

File: lib/Controller/MarketController.php [hunk 2]
```diff
@@ -328,6 +346,41 @@ public function requestDemoLicenseKeyFromMarket() {
...
+
+		return new DataResponse([
+			'loginUrl' => "$appStoreUrl/api-login?callbackUrl=$callbackUrl&codeChallenge=$codeChallenge"
+		]);
+	}
+
+	/**
+	 * @NoCSRFRequired
+	 */
+	public function receiveMarketplaceLoginToken($token) {
```
Comment (VicDeo) [inline lib/Controller/MarketController.php]: PHPDOC for `$token` is missing

File: lib/HttpService.php [hunk 3]
```diff
@@ -141,6 +142,19 @@ public function validateKey($apiKey) {
 		return $this->httpGet($url, [], $apiKey);
 	}
 
+	public function exchangeLoginTokenForApiKey($loginToken, $codeVerifier) {
```
Comment (VicDeo) [inline lib/HttpService.php]: PHPDoc is missing

File: lib/HttpService.php [hunk 4]
```diff
@@ -273,6 +286,34 @@ private function httpGet($path, $options, $apiKey) {
 		return $response;
 	}
 
+	private function httpPost($path, $options) {
```
Comment (VicDeo) [inline lib/HttpService.php]: PHPDoc

**Comment by IljaN** [comment]
@VicDeo Adressed your changes. Tough not sure about: 
> Vue templates are already indented with spaces so the formatting looks weird for them.
Care to explain?

**Comment by IljaN** [comment]
Please don't merge yet

File: src/App.vue [hunk 5]
```diff
@@ -17,6 +18,7 @@
     import Mixins from './mixins.js'
     import Navigation from './components/Navigation.vue'
     import Trial from './components/Trial.vue'
+	import Login from './components/Login.vue'
```
Comment (VicDeo) [inline src/App.vue]: @IljaN indented with tab

**Comment by VicDeo** [comment]
@IljaN open the diff for src/App.vue here and you see that your changes are not aligned with the rest in this file. I spotted out one line, the rest are same (tabs instead of spaces)

**Comment by IljaN** [comment]
Hmhh App.vue seems to be mixed, sometimes tabs, sometimes spaces :see_no_evil:

**Comment by IljaN** [comment]
@VicDeo Please re-review

File: lib/HttpService.php [hunk 6]
```diff
@@ -141,6 +142,28 @@ public function validateKey($apiKey) {
 		return $this->httpGet($url, [], $apiKey);
 	}
 
+	/**
+	 *
+	 * Exchange login token for api key
+	 *
+	 * @param $loginToken
```
Comment (VicDeo) [inline lib/HttpService.php]: type is missing

File: lib/HttpService.php [hunk 7]
```diff
@@ -141,6 +142,28 @@ public function validateKey($apiKey) {
...
+	 * @param $codeVerifier
```
Comment (VicDeo) [inline lib/HttpService.php]: type is missing

File: lib/HttpService.php [hunk 8]
```diff
@@ -273,6 +295,40 @@ private function httpGet($path, $options, $apiKey) {
 		return $response;
 	}
 
+	/**
+	 * @param $path
```
Comment (VicDeo) [inline lib/HttpService.php]: type is missing

File: lib/HttpService.php [hunk 9]
```diff
@@ -273,6 +295,40 @@ private function httpGet($path, $options, $apiKey) {
...
+	 * @param $options
```
Comment (VicDeo) [inline lib/HttpService.php]: type is missing

**Pull request merged**